pkg/apis/kafka/v1alpha1/reset_offset_annotations.go

```diff
@@ -0,0 +1,87 @@
+/*
+Copyright 2021 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"encoding/json"
+)
+
+const (
+	ResetOffsetAnnotationTopic      = "topic"
+	ResetOffsetAnnotationGroup      = "group"
+	ResetOffsetAnnotationPartitions = "partitions"
+)
+
+// OffsetMapping is a simple struct representing a single Kafka
+// Topic/Partition Offset value before and after repositioning
+type OffsetMapping struct {
+	Partition int32 `json:"partition"`
+	OldOffset int64 `json:"oldOffset"`
+	NewOffset int64 `json:"newOffset"`
```

Comment:
I'm sorry, I don't understand the reason of this. Is it for some sort of "signaling" the result? If that's the case, why it's not in the `ResetOffsetStatus`?

Reply:
The struct is used to track the old/new offsets and is written to the duck Status Annotations which seems to be their intended use right...

```
        // Annotations is additional Status fields for the Resource to save some
	// additional State as well as convey more information to the user. This is
	// roughly akin to Annotations on any k8s resource, just the reconciler conveying
	// richer information outwards.
	Annotations map[string]string `json:"annotations,omitempty"`
```

I think this makes sense vs trying to inline this data in the Status block itself.  I can rework it if I've misunderstood the intent though?

Reply:
As to the "why"... I think knowing the old/new offsets is really useful historical information for the end user to save off and track when they do this kind of thing (recovering from downtime, etc).   At a minimum it doesn't hurt anything to expose it ; )

Reply:
Yeah I think this definitely goes in the CRD status, not inside the conditions but just as a field of the status. `Annotations` are a bit hacky for that use case IMO

Reply:
It IS inside the Status section as Annotations (not inside Conditions)...
```yaml
  status:
    annotations:
      group: my-group
      partitions: '[{"partition":0,"oldOffset":1,"newOffset":2},{"partition":1,"oldOffset":2,"newOffset":3},{"partition":2,"oldOffset":3,"newOffset":4},{"partition":3,"oldOffset":4,"newOffset":5}]'
      topic: my-topic
    conditions:
    - lastTransitionTime: "2021-05-03T16:55:21Z"
      status: Unknown
      type: ConsumerGroupsStarted
```
...which appears to be the intended use of the duck/v1 Status type?   I think if we are not going to use that type in this use case I would argue it should be removed from the duck type altogether because what else would it possibly be intended for?

Reply:
> which appears to be the intended use of the duck/v1 Status type?

TBH, I don't know what's the best solution here... Maybe should we mimic what job does?

```
type JobStatus struct {
	Conditions []JobCondition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type" protobuf:"bytes,1,rep,name=conditions"`

	StartTime *metav1.Time `json:"startTime,omitempty" protobuf:"bytes,2,opt,name=startTime"`

	CompletionTime *metav1.Time `json:"completionTime,omitempty" protobuf:"bytes,3,opt,name=completionTime"`

	Active int32 `json:"active,omitempty" protobuf:"varint,4,opt,name=active"`

	Succeeded int32 `json:"succeeded,omitempty" protobuf:"varint,5,opt,name=succeeded"`

	Failed int32 `json:"failed,omitempty" protobuf:"varint,6,opt,name=failed"`
}
```

`Conditions` is the same as our conditions stuff inside `duckv1.Status`, while the rest of the fields, to me, sounds like what you're pushing inside annotations. So I fail to see the reason why we should have that information inside `annotations`.

We can't remove `duckv1.Status` though, because it contains `ObservedGeneration`. That field is necessary to make our reconcilers work. It also contains `Conditions`, which is something we need too :)

Reply:
The question is... is that `optional` ... or more something "real":
https://github.com/knative/pkg/blob/main/apis/duck/v1/status_types.go#L50-L54

I could live with both... pumping/stashing these on the `Annotations`, or more explicit :man_shrugging:

Reply:
From reading the duckv1 Status.Annotations documentation I think we are using that field exactly as intended and don't see a problem.  I can move the fields to be explicit top level members of `ResetOffsetStatus` if there is consensus but... I would then argue that the duckv1 Status.Annotations field is worthless.   What kind of information WOULD you put there if not this?

Reply:
I guess the argument is that the Status.Annotations field is not "official" enough or something?

Reply:
I will make the change (moving to top level status fields) - I was just trying to use the duck types as I interpreted the intent from the docs.

Reply:
> I would then argue that the duckv1 Status.Annotations field is worthless. What kind of information WOULD you put there if not this?

TBH I have no strong argument about that and I frankly never used `Status.Annotation` for something, nor I'm aware of any API eventing using it. I won't argue its usefulness or not, because I don't have enough context, but keeping things in the `ResetOffsetStatus` struct IMO is better and you can have nested informations without serializing to json (like you do for partitions). If you look at how kafka source status and kafka channel status show some of their status metadata, they don't use annotations.

Reply:
sure, fair enough - most of the discussion was just confusion on my part as to what was being requested and what the concern was.  thx

Reply:
OK, so I refactored and it now looks like this...
```yaml
status:
  conditions:
  - lastTransitionTime: "2021-05-05T20:37:09Z"
    message: failed to start consumergroup 123
    reason: FailedToStartConsumerGroup
    status: "False"
    type: ConsumerGroupsStarted
  - lastTransitionTime: "2021-05-05T20:37:09Z"
    status: "True"
    type: ConsumerGroupsStopped
  - lastTransitionTime: "2021-05-05T20:37:09Z"
    status: "True"
    type: OffsetsUpdated
  - lastTransitionTime: "2021-05-05T20:37:09Z"
    status: "True"
    type: ResetInitiated
  - lastTransitionTime: "2021-05-05T20:37:09Z"
    message: failed to start consumergroup 123
    reason: FailedToStartConsumerGroup
    status: "False"
    type: Succeeded
  group: my-group
  observedGeneration: 1
  partitions:
  - newOffset: 2
    oldOffset: 1
    partition: 0
  - newOffset: 3
    oldOffset: 2
    partition: 1
  topic: my-topic
```

Reply:
thanks for the refactors, @travis-minke-sap 

/lgtm